### PR TITLE
Common: Support PLDM sensor monitor

### PR DESCRIPTION
**Pull request by BonnieLo**
# Description
- Implement PLDM sensor reading.

# Motivation
- Support PLDM sensor monitor.

# Test plan
- Build code: Pass
- Check PLDM sensor monitor thread: Pass
- Using LFBMC get PLDM info: Pass

# Log
1. Check PLDM sensor monitor is normal. uart:~$ kernel stacks
0x40200 mctptx_01_06_40 (real size 2048):       unused 1708     usage 340 / 2048 (16 %)
0x40138 mctprx_01_06_40 (real size 4096):       unused 2900     usage 1196 / 4096 (29 %)
0x5cc68 USB_handler (real size 2048):   unused 1748     usage 300 / 2048 (14 %)
0x59628 IPMI_thread (real size 4096):   unused 3236     usage 860 / 4096 (20 %)
0x5bad8 DIMM_PLDM_SENSOR_THREAD (real size 3056):       unused 2356     usage 700 / 3056 (22 %)
0x5ba10 CPU_PLDM_SENSOR_THREAD (real size 3056):        unused 2804     usage 252 / 3056 (8 %)
0x5b948 ADC_PLDM_SENSOR_THREAD (real size 3056):        unused 2676     usage 380 / 3056 (12 %)
0x58398 WDT_thread (real size 256):     unused 44       usage 212 / 256 (82 %)
0x5a7b8 pldm_wait_resp_to (real size 1024):     unused 812      usage 212 / 1024 (20 %)
0x5a6f0 monitor_tid (real size 1024):   unused 812      usage 212 / 1024 (20 %)
0x5ce30 shell_uart (real size 2048):    unused 1052     usage 996 / 2048 (48 %)
0x58278 gpio_workq (real size 3072):    unused 2780     usage 292 / 3072 (9 %)
0x3b9f0 ADC0       (real size 512):     unused 252      usage 260 / 512 (50 %)
0x3bdc8 ADC1       (real size 512):     unused 268      usage 244 / 512 (47 %)
0x653b8 sysworkq   (real size 2048):    unused 1788     usage 260 / 2048 (12 %)
0x5cfe0 usbdworkq  (real size 1024):    unused 556      usage 468 / 1024 (45 %)
0x5cef8 usbworkq   (real size 1024):    unused 764      usage 260 / 1024 (25 %)
0x5cd68 logging    (real size 768):     unused 260      usage 508 / 768 (66 %)
0x5d218 idle 00    (real size 512):     unused 460      usage 52 / 512 (10 %)
0x73f58 IRQ 00     (real size 2048):    unused 1544     usage 504 / 2048 (24 %)

2. Get sensor and sensor reading on LF BMC. root@bmc:~# busctl tree xyz.openbmc_project.PLDM
`- /xyz
  `- /xyz/openbmc_project
    |- /xyz/openbmc_project/inventory
    | `- /xyz/openbmc_project/inventory/Item
    |   `- /xyz/openbmc_project/inventory/Item/Board
    |     `- /xyz/openbmc_project/inventory/Item/Board/PLDM_Device_1
    |- /xyz/openbmc_project/pldm
    `- /xyz/openbmc_project/sensors
      |- /xyz/openbmc_project/sensors/power
      | |- /xyz/openbmc_project/sensors/power/PLDM_Device_46_1
      | |- /xyz/openbmc_project/sensors/power/PLDM_Device_47_1
      | |- /xyz/openbmc_project/sensors/power/PLDM_Device_48_1
      | `- /xyz/openbmc_project/sensors/power/PLDM_Device_56_1
      |- /xyz/openbmc_project/sensors/temperature
      | |- /xyz/openbmc_project/sensors/temperature/PLDM_Device_14_1
      | |- /xyz/openbmc_project/sensors/temperature/PLDM_Device_4_1
      | |- /xyz/openbmc_project/sensors/temperature/PLDM_Device_5_1
      | `- /xyz/openbmc_project/sensors/temperature/PLDM_Device_6_1
      `- /xyz/openbmc_project/sensors/voltage
        |- /xyz/openbmc_project/sensors/voltage/PLDM_Device_24_1
        `- /xyz/openbmc_project/sensors/voltage/PLDM_Device_26_1

- ADC 3.3V (test on HD system) root@bmc:~# busctl introspect xyz.openbmc_project.PLDM /xyz/openbmc_project/sensors/voltage/PLDM_Device_26_1
NAME                                                  TYPE      SIGNATURE RESULT/VALUE                             FLAGS
org.freedesktop.DBus.Introspectable                   interface -         -                                        -
.Introspect                                           method    -         s                                        -
org.freedesktop.DBus.Peer                             interface -         -                                        -
.GetMachineId                                         method    -         s                                        -
.Ping                                                 method    -         -                                        -
org.freedesktop.DBus.Properties                       interface -         -                                        -
.Get                                                  method    ss        v                                        -
.GetAll                                               method    s         a{sv}                                    -
.Set                                                  method    ssv       -                                        -
.PropertiesChanged                                    signal    sa{sv}as  -                                        -
xyz.openbmc_project.Association.Definitions           interface -         -                                        -
.Associations                                         property  a(sss)    1 "chassis" "all_sensors" "/xyz/openb... emits-change writable
xyz.openbmc_project.Sensor.Threshold.Critical         interface -         -                                        -
.CriticalAlarmHigh                                    property  b         false                                    emits-change writable
.CriticalAlarmLow                                     property  b         false                                    emits-change writable
.CriticalHigh                                         property  d         3.56985                                  emits-change writable
.CriticalLow                                          property  d         3.04095                                  emits-change writable
.CriticalHighAlarmAsserted                            signal    d         -                                        -
.CriticalHighAlarmDeasserted                          signal    d         -                                        -
.CriticalLowAlarmAsserted                             signal    d         -                                        -
.CriticalLowAlarmDeasserted                           signal    d         -                                        -
xyz.openbmc_project.Sensor.Threshold.Warning          interface -         -                                        -
.WarningAlarmHigh                                     property  b         false                                    emits-change writable
.WarningAlarmLow                                      property  b         false                                    emits-change writable
.WarningHigh                                          property  d         3.5343                                   emits-change writable
.WarningLow                                           property  d         3.0723                                   emits-change writable
.WarningHighAlarmAsserted                             signal    d         -                                        -
.WarningHighAlarmDeasserted                           signal    d         -                                        -
.WarningLowAlarmAsserted                              signal    d         -                                        -
.WarningLowAlarmDeasserted                            signal    d         -                                        -
xyz.openbmc_project.Sensor.Value                      interface -         -                                        -
.MaxValue                                             property  d         3.465                                    emits-change writable
.MinValue                                             property  d         3.135                                    emits-change writable
.Unit                                                 property  s         "xyz.openbmc_project.Sensor.Value.Uni... emits-change writable
.Value                                                property  d         3.334                                    emits-change writable
xyz.openbmc_project.State.Decorator.Availability      interface -         -                                        -
.Available                                            property  b         true                                     emits-change writable
xyz.openbmc_project.State.Decorator.OperationalStatus interface -         -                                        -
.Functional                                           property  b         true                                     emits-change writable

- ADC 12V(test on HD system) root@bmc:~# busctl introspect xyz.openbmc_project.PLDM /xyz/openbmc_project/sensors/voltage/PLDM_Device_24_1
NAME                                                  TYPE      SIGNATURE RESULT/VALUE                             FLAGS
org.freedesktop.DBus.Introspectable                   interface -         -                                        -
.Introspect                                           method    -         s                                        -
org.freedesktop.DBus.Peer                             interface -         -                                        -
.GetMachineId                                         method    -         s                                        -
.Ping                                                 method    -         -                                        -
org.freedesktop.DBus.Properties                       interface -         -                                        -
.Get                                                  method    ss        v                                        -
.GetAll                                               method    s         a{sv}                                    -
.Set                                                  method    ssv       -                                        -
.PropertiesChanged                                    signal    sa{sv}as  -                                        -
xyz.openbmc_project.Association.Definitions           interface -         -                                        -
.Associations                                         property  a(sss)    1 "chassis" "all_sensors" "/xyz/openb... emits-change writable
xyz.openbmc_project.Sensor.Threshold.Critical         interface -         -                                        -
.CriticalAlarmHigh                                    property  b         false                                    emits-change writable
.CriticalAlarmLow                                     property  b         false                                    emits-change writable
.CriticalHigh                                         property  d         13.3488                                  emits-change writable
.CriticalLow                                          property  d         10.7088                                  emits-change writable
.CriticalHighAlarmAsserted                            signal    d         -                                        -
.CriticalHighAlarmDeasserted                          signal    d         -                                        -
.CriticalLowAlarmAsserted                             signal    d         -                                        -
.CriticalLowAlarmDeasserted                           signal    d         -                                        -
xyz.openbmc_project.Sensor.Threshold.Warning          interface -         -                                        -
.WarningAlarmHigh                                     property  b         false                                    emits-change writable
.WarningAlarmLow                                      property  b         false                                    emits-change writable
.WarningHigh                                          property  d         13.2192                                  emits-change writable
.WarningLow                                           property  d         10.8192                                  emits-change writable
.WarningHighAlarmAsserted                             signal    d         -                                        -
.WarningHighAlarmDeasserted                           signal    d         -                                        -
.WarningLowAlarmAsserted                              signal    d         -                                        -
.WarningLowAlarmDeasserted                            signal    d         -                                        -
xyz.openbmc_project.Sensor.Value                      interface -         -                                        -
.MaxValue                                             property  d         12.96                                    emits-change writable
.MinValue                                             property  d         11.04                                    emits-change writable
.Unit                                                 property  s         "xyz.openbmc_project.Sensor.Value.Uni... emits-change writable
.Value                                                property  d         12.097                                   emits-change writable
xyz.openbmc_project.State.Decorator.Availability      interface -         -                                        -
.Available                                            property  b         true                                     emits-change writable
xyz.openbmc_project.State.Decorator.OperationalStatus interface -         -                                        -
.Functional                                           property  b         true                                     emits-change writable

### CODE REVIEW

File: common/service/sensor/pldm_sensor.c [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+		return;
+	}
+
+	pldm_sensor_list[thread_id] = plat_pldm_sensor_load(thread_id);
+	if (pldm_sensor_list[thread_id] == NULL) {
+		LOG_ERR("Failed to load PLDM sensor list of thread%d, ret%d", thread_id, ret);
+		return;
+	}
+
+	for (sensor_num = 0; sensor_num < pldm_sensor_count; sensor_num++) {
```
Comment (amithash) [inline common/service/sensor/pldm_sensor.c:238]: This is starting to get harder to read. Can you move the body if this loop into a function of its own?
That way the indentation is not all the way to the right and it gets harder to know where things start.
That way you can pass it &pldm_sensor_list[thread_id][sensor_num] and that function would be so much more easier to read.

File: common/service/sensor/pldm_sensor.c [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+					     NULL)) {
+				pldm_sensor_list[thread_id][sensor_num]
+					.pldm_sensor_cfg.cache_status = PLDM_SENSOR_FAILED;
+				continue;
+			}
+		}
+	}
+
+	while (1) {
+		for (sensor_num = 0; sensor_num < pldm_sensor_count; sensor_num++) {
```
Comment (amithash) [inline common/service/sensor/pldm_sensor.c:281]: Same thing for this for-loop. Move the body into its own function.
You could pass it &pldm_sensor_list[thread_id][sensor_num].

File: common/service/sensor/pldm_sensor.c [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+	}
+
+	if (resolution == 0) {
+		// The value of resolution couldn't be 0
+		return PLDM_ERROR_INVALID_DATA;
+	}
+
+	// Convert two byte integer, two byte decimal sensor format to float
+	sensor_reading_byte = malloc(sizeof(int));
+	memcpy(sensor_reading_byte, &cache_reading, sizeof(int));
```
Comment (amithash) [inline common/service/sensor/pldm_sensor.c:129]: Why allocate? Theoritically, you could have gotten away with `uint8_t sensor_reading_byte[sizeof(int)];`
But, assuming we are in little-endian which this is assuming
integer = cache_reading & 0xffff;
decimal = (float)(cache_reading >> 16) / 1000.0;

But I am assuming little endian. I have a feeling this arch is big-endian.

Comment (amithash) [reply common/service/sensor/pldm_sensor.c:129]: If the byte order is off, use a builtin byte swap function: https://docs.zephyrproject.org/latest/doxygen/html/sys_2byteorder_8h.html

**Review by amithash — CHANGES_REQUESTED** [review]
Some readability and perf suggestions.

File: common/service/sensor/pldm_sensor.c (hunk 1 above)
Comment (RickyWu-wiwynn) [reply common/service/sensor/pldm_sensor.c:293]: Modified with your suggestions. Thanks.

File: common/service/sensor/pldm_sensor.c (hunk 2 above)
Comment (RickyWu-wiwynn) [reply common/service/sensor/pldm_sensor.c:300]: Done.

File: common/service/sensor/pldm_sensor.c (hunk 3 above)
Comment (RickyWu-wiwynn) [reply common/service/sensor/pldm_sensor.c]: Done.

**Pull request closed without merging**